### ISSUE TITLE
edge channel stricter about `build-environment` types in parts

### ISSUE THREAD

**Issue by lengau** [canonical/snapcraft#4978]
### Bug Description

The `edge` channel can no longer build charmcraft due to a value error in `parts.spread.build-environment`

### To Reproduce

1. Clone charrmcraft (I'm on commit `0d3b27136fb9c21e6d7c091dc4ca4ac6bfee3aca`)
2. `snapcraft pack` with snapcraft revision 12433 (`8.3.2.post34+git655e543a`)

### Environment

KDE Neon noble

### snapcraft.yaml

```shell
https://github.com/canonical/charmcraft/blob/0d3b27136fb9c21e6d7c091dc4ca4ac6bfee3aca/snap/snapcraft.yaml#L187
```

### Relevant log output

```shell
lengau@ratel:~/Work/Code/charmcraft-2$ snapcraft pack
Bad snapcraft.yaml content:
- Value error, value must be a list of Annotated: [{'CGO_ENABLED': 0}] (in field 'parts.spread.build-environment')                                                              
Full execution log: '/home/lengau/.local/state/snapcraft/log/snapcraft-20240815-165625.696525.log'
```

### Additional context

Changing the value of `CGO_ENABLED` from `0` to `"0"`resolved it

**Comment by mr-cal** [comment]
This likely broke due to pydantic 2 dropping this behavior by default: https://docs.pydantic.dev/dev/api/config/#pydantic.config.ConfigDict.coerce_numbers_to_str

Snapcraft probably needs fixes in places like this: https://github.com/canonical/snapcraft/blob/037c230170cfa33760c27cbefd4388a78b148294/snapcraft/models/project.py#L945

Craft-parts probably needs fixes too: https://github.com/canonical/craft-parts/blob/7ea8655b596b359e8542d171bea011f672e1c996/craft_parts/parts.py#L67

**Comment by lengau** [comment]
Thanks @mr-cal  - I've made an upstream PR for it. Once that's merged an update will fix this.

**Comment by mr-cal** [comment]
Fixed via #5051
